### PR TITLE
Add eval_temp option and ability to log extra reward metrics by returning dict

### PR DESCRIPTION
**Pull request by tmfs10**
- **added extra rm logging for wandb and option for eval temperature**
- **debugging**
- **debugged**
- **debugged eval reward metrics**

### CODE REVIEW

File: openrlhf/trainer/ppo_utils/experience_maker.py [hunk 1]
```diff
@@ -342,11 +342,17 @@ def make_experience(self, extra_rm_args, samples: Samples) -> Experience:
...
         info = {
             "kl": masked_mean(kl, action_mask, dim=-1),
-            "reward": r,
             "response_length": samples.response_length,
             "total_length": samples.total_length,
             "num_actions": num_actions,
         }
+
+        if type(r) == dict:
+            assert 'reward' in r
```
Comment (gwarmstrong) [inline openrlhf/trainer/ppo_utils/experience_maker.py:351]: what does this assertion accomplish? under what conditions does this not hold?

File: openrlhf/trainer/ray/ppo_actor.py [hunk 2]
```diff
@@ -380,7 +380,11 @@ def custom_collate_fn(batch):
                 eval_data = eval_data.select(range(min(samples_in_limited_batches, len(eval_data))))
 
             self.eval_dataset = PromptDataset(
-                eval_data, self.tokenizer, strategy, input_template=args.input_template
+                eval_data, 
+                self.tokenizer, 
+                strategy, 
+                input_template=args.input_template,
+                max_prompt_length=args.prompt_max_len if not args.truncated_prompts else None,
```
Comment (gwarmstrong) [inline openrlhf/trainer/ray/ppo_actor.py:387]: can you add a `help` description to the argument here: https://github.com/Kipok/OpenRLHF/blob/97bccd79b66c4b975af258839ecbd1edd486941a/openrlhf/cli/train_ppo_ray.py#L378
describing that `--no_truncated_prompts` filters out prompts that would have been truncated?

File: openrlhf/trainer/ppo_utils/experience_maker.py (hunk 1 above)
Comment (tmfs10) [reply openrlhf/trainer/ppo_utils/experience_maker.py:351]: It should always hold. I just want to make sure that 'reward' key is present if a dict is being returned from the reward function

File: openrlhf/trainer/ray/ppo_actor.py (hunk 2 above)
Comment (tmfs10) [reply openrlhf/trainer/ray/ppo_actor.py:387]: Added

File: openrlhf/trainer/ppo_utils/experience_maker.py (hunk 1 above)
Comment (gwarmstrong) [reply openrlhf/trainer/ppo_utils/experience_maker.py:351]: So if you don't expect to see this violated, why check here? If it is violated, what's the consequence on the following code?

Trying to understand why the check is need. If it isn't, I would remove it.

**Comment by tmfs10** [comment]
This is causing an issue with eval right now so please hold off on the merge

File: openrlhf/trainer/ppo_utils/experience_maker.py (hunk 1 above)
Comment (tmfs10) [reply openrlhf/trainer/ppo_utils/experience_maker.py:351]: If the reward key is not returned from the external reward function, then the entire RL loop will break. Having an assert here makes it easier for a user to debug if they accidentally forgot to return the reward

**Comment by tmfs10** [comment]
All bugs fixed. @Kipok can you review and merge?

File: openrlhf/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -498,30 +520,37 @@ def training_step_critic(self, experience: Experience) -> Dict[str, float]:
...
-            rewards = torch.stack(eval_buffer['reward'])
-            rewards = self.strategy.all_gather(rewards)
-            logs_dict['reward'] = rewards.mean().item()
+        for k, v in eval_buffer.items():
+            if len(eval_buffer[k]) > 0:
+                metrics = torch.cat(eval_buffer[k])
+                logs_dict[k] = metrics.mean().item()
 
-            return logs_dict
+        print('Evaluation:', logs_dict)
```
Comment (gwarmstrong) [inline openrlhf/trainer/ppo_trainer.py]:
```suggestion
        self.strategy.print('Evaluation:', logs_dict)
```

**Review by gwarmstrong — COMMENTED** [review]
@tmfs10 can you please write a general description of the logic that was introduced w.r.t reward and metric prefixes? What is the intent? What behavior does it introduce/break and which modules does it affect?

**Comment by tmfs10** [comment]
> @tmfs10 can you please write a general description of the logic that was introduced w.r.t reward and metric prefixes? What is the intent? What behavior does it introduce/break and which modules does it affect?

The intent is to add the ability to track additional metrics an external reward function might compute on top of the reward itself. In particular, for code generation, while our reward is usually fraction of unit tests passed, it's also very useful for us to track the number of problems which pass all unit tests which the reward function also computes but is impossible to record in the existing setup.

Here, if the user wants to record additional metrics, instead of returning a reward tensor, they return a dict where the 'reward' key has the reward tensor, and any keys with the prefix 'metric_' are logged into wandb separately.

However the existing workflow is undisturbed -- if the user just returns a reward tensor, that's handled as before.

Separately, I added the --eval_temperature option to be able to change the evaluation temperature instead of it being always fixed to 0. It defaults to 0.

**Review by gwarmstrong — APPROVED** [review]
Thanks @tmfs10 for the clarification.
